doc/recipe/preprocessor.rst

```diff
@@ -180,12 +152,129 @@ steps:
+:ref:`weighting_landsea_fraction<land/sea fraction weighting>` ``sftlf``, ``sftof``           land_area_fraction, sea_area_fraction
+============================================================== ============================== =====================================
+
+Only one of the listed variables is required. Ancillary variables can be defined
+in the recipe as described in :ref:`ancillary_variables`. In some cases,
+preprocessor functions may work without ancillary variables, this is documented
+case by case in the preprocessor function definition.
+
+By default, ancillary variables and cell measures will be removed from the
+variable before saving it to file because they can be as big as the main
+variable.
+To keep the ancillary variables and cell measures, disable the preprocessor
+function :func:`esmvalcore.preprocessor.remove_ancillary_variables` that removes
+them by setting ``remove_ancillary_variables: false`` in the preprocessor
+in the recipe.
+
+Examples
+--------
+
+Compute the global mean surface air temperature, while
+:ref:`automatically selecting the best matching ancillary dataset <ancillary_dataset_wildcards>`:
+
+.. code-block:: yaml
 
-Fx variables as cell measures or ancillary variables
-====================================================
-The following preprocessors may require the use of ``fx_variables`` to be able
-to perform the computations:
+  datasets:
+    - dataset: BCC-ESM1
+      project: CMIP6
+      ensemble: r1i1p1f1
+      grid: gn
+    - dataset: CanESM2
+      project: CMIP5
+      ensemble: r1i1p1,
+
+  preprocessors:
+    global_mean:
+      area_statistics:
+        operator: mean
+
+  diagnostics:
+    example_diagnostic:
+      description: Global mean temperature.
+      variables:
+        tas:
+          mip: Amon
+          preprocessor: global_mean
+          exp: historical
+          timerange: '1990/2000'
+          ancillary_variables:
+            - short_name: areacella
+              mip: fx
+              exp: '*'
+              activity: '*'
+              ensemble: '*'
+      scripts: null
+
+Attach the land area fraction as an ancillary variable to surface air
+temperature and store both in the same file:
+
+.. code-block:: yaml
+
+  datasets:
+    - dataset: BCC-ESM1
+      ensemble: r1i1p1f1
+      grid: gn
+
+  preprocessors:
+    keep_land_area_fraction:
+      remove_ancillary_variables: false
+
+  diagnostics:
+    example_diagnostic:
+      description: Attach land area fraction.
+      variables:
+        tas:
+          mip: Amon
+          project: CMIP6
+          preprocessor: keep_land_area_fraction
+          exp: historical
+          timerange: '1990/2000'
+          ancillary_variables:
+            - short_name: sftlf
+              mip: fx
+              exp: 1pctCO2
+      scripts: null
+
+.. _`Fx variables as cell measures or ancillary variables`:
+
+Legacy method of specifying ancillary variables
```

Comment:
The new syntax helps in specifying ancillary variables when they are not saved where expected, but there are many instances in which they are where they should be and the tool was able to load them just be defining the right preprocessor.

I think it's fine to give the possibility to define ancillary variables at a dataset level, but I don't think that means that we should completely get rid of the possibility of defining a recipe just like this:

```yaml
  datasets:
    - dataset: BCC-ESM1
      ensemble: r1i1p1f1
      grid: gn

  preprocessors:
    global_mean:
      area_statistics:
        operator: 'mean'

  diagnostics:
    example_diagnostic:
      description: Attach land area fraction.
      variables:
        tas:
          mip: Amon
          project: CMIP6
          preprocessor: global_mean
          exp: historical
          timerange: '1990/2000'
      scripts: null

```

For what I have seen, people tend to have a hard time understanding what is an ancillary variable or a cell measure, and why are they needed in certain preprocessor functions.  So I think it's a bit too much to force them to always specify ancillary_variables, when there are many recipes that were working without having to specify them.

Reply:
Good point, I'll see if I can do something to make it more automatic.

For `areacella`, `areacello`, `sftlf`, and `sftof` this seems straightforward because they are only available in a single `mip`. For `volcello` and `sftgif`the situation is more complicated. At the moment it is not possible to specify the `mip` as a `*`, it would be nice if this could work for ancillary variables, but for ancillary variables that are available in multiple `mip`s that can only work if there is some clear way of picking the correct ancillary `mip` based on the `mip` of the main variable.

Reply:
I feel that matching the facets would be nicer, but I'm not sure if it's possible. Discarding things that cannot be broadcasted to the right shape means that
1) you need to download all the files, even if you're never going to use some of them
2) ancillary variables that need a fix might be silently discarded
2) it might work when it shouldn't but the arrays coincidentally have a compatible shape

Reply:
Then maybe the matching could be done during the copy of the facets, by importing the CMOR TABLES. Check if the ancillary exists in `fx` and if it does not, use the mip from the variable. So something like:

```python
        for ancillary in self.ancillaries:
            # The short_name and mip of the ancillary variable are probably
            # different from the main variable, so don't copy those facets.
            skip = ('short_name',  'mip')
            ancillary_facets = {
                k: v
                for k, v in facets.items() if k not in skip
            }
       
         
            new_ancillary = ancillary.copy(**ancillary_facets)
            fx = CMOR_TABLES[project].get_variable('fx', short_name)
            if not fx:
                new_ancillary['mip'] = self.facets['mip']
            new.ancillaries.append(new_ancillary)

```
You could even check if volcello and sftgif exist in the mip of the variable, and if they do not, raise an error earlier at this point.

Reply:
I implemented support for automatically adding ancillaries in [fa8f392](https://github.com/ESMValGroup/ESMValCore/pull/1609/commits/fa8f3921ac63f4ea831e693f4805443a3fe1a9a9). Would you have time to test if it works as expected? To use it, run the tool with `--use-legacy-ancillaries=False`.

Reply:
Yes, on it. Thanks!!

Reply:
It seems that the CMOR information is getting loaded before the matching:
```
2023-02-08 15:23:32,660 UTC [923534] DEBUG   Populating list of datasets for variable tas in diagnostic example_diagnostic
2023-02-08 15:23:32,661 UTC [923534] DEBUG   Expanding dataset globs for dataset Dataset: tas, Amon, CMIP6, EC-Earth3, historical, r1i1p1f1, gr, ancillaries: areacella, *, *, *, *, this may take a while..
2023-02-08 15:23:32,663 UTC [923534] DEBUG   Looking for files matching:
/scratch/b/b381943/climate_data/CMIP6/C2023-02-08 15:23:32,660 UTC [923534] DEBUG   Populating list of datasets for variable tas in diagnostic example_diagnostic
2023-02-08 15:23:32,661 UTC [923534] DEBUG   Expanding dataset globs for dataset Dataset: tas, Amon, CMIP6, EC-Earth3, historical, r1i1p1f1, gr, ancillaries: areacella, *, *, *, *, this may take a while..
2023-02-08 15:23:32,663 UTC [923534] DEBUG   Looking for files matching:
/scratch/b/b381943/climate_data/CMIP6/CMIP/*/EC-Earth3/historical/r1i1p1f1/Amon/tas/gr/*/tas_Amon_EC-Earth3_historical_r1i1p1f1_gr*.nc
2023-02-08 15:23:33,011 UTC [923534] INFO    Maximum memory used (estimate): 0.0 GB
2023-02-08 15:23:33,012 UTC [923534] INFO    Sampled every second. It may be inaccurate if short but high spikes in memory consumption occur.
2023-02-08 15:23:33,014 UTC [923534] ERROR   Unable to load CMOR table (project) 'CMIP6' for variable 'areacella' with mip '*'MIP/*/EC-Earth3/historical/r1i1p1f1/Amon/tas/gr/*/tas_Amon_EC-Earth3_historical_r1i1p1f1_gr*.nc
2023-02-08 15:23:33,011 UTC [923534] INFO    Maximum memory used (estimate): 0.0 GB
2023-02-08 15:23:33,012 UTC [923534] INFO    Sampled every second. It may be inaccurate if short but high spikes in memory consumption occur.
2023-02-08 15:23:33,014 UTC [923534] ERROR   Unable to load CMOR table (project) 'CMIP6' for variable 'areacella' with mip '*'
```
I tried a recipe like the one above

Reply:
Thanks for testing! Could you share the recipe? I tried with
```yaml
# ESMValTool
# recipe_python.yml
---
documentation:
  description: |
    Example recipe that plots a map and timeseries of temperature.

  title: Recipe that runs an example diagnostic written in Python.

  authors:
    - andela_bouwe
    - righi_mattia

  maintainer:
    - schlund_manuel

  references:
    - acknow_project

  projects:
    - esmval
    - c3s-magic

datasets:
  - dataset: EC-Earth3
    ensemble: r1i1p1f1
    grid: gr

preprocessors:
  global_mean:
    area_statistics:
      operator: 'mean'

diagnostics:
  example_diagnostic:
    description: Attach land area fraction.
    variables:
      tas:
        mip: Amon
        project: CMIP6
        preprocessor: global_mean
        exp: historical
        timerange: '1990/2000'
    scripts: null
```
which I ran with `esmvaltool run recipe_python.yml --use-legacy-ancillaries=False`
and that seems to work fine:
```
2023-02-08 21:02:42,667 UTC [245787] INFO    Found input files for Dataset: tas, Amon, CMIP6, EC-Earth3, CMIP, historical, r1i1p1f1, gr, v20200310, ancillaries: areacella, fx, v20210324
```

Reply:
I think mine failed because it had a wildcard in the ensemble. If the ancillaries are automatically guessed, 
 then the recipe cannot use wildcards for any tag?